utils/trie_writer.hh

```diff
@@ -0,0 +1,1124 @@
+        ptr() {}
+        [[nodiscard]] T* get() const noexcept {
+            if constexpr (SANITIZE_BUMP_ALLOCATOR) {
+                if (_ptr) {
+                    auto legal_timestamp_it = std::prev(this->_alctr->_max_legal_timestamp.lower_bound(this->_global_pos + this->_size - 1));
+                    EXPENSIVE_ASSERT(this->discard_timestamp >= legal_timestamp_it->second);
+                    EXPENSIVE_ASSERT(this->_global_pos + (this->_size != 0) <= this->_alctr->global_pos());
+                }
+            }
+            return _ptr;
+        }
+        [[nodiscard]] T& operator[] (size_t i) const noexcept {
+            #if SANITIZE_BUMP_ALLOCATOR
+                EXPENSIVE_ASSERT(_size);
+            #endif
+            return get()[i];
+        };
+        [[nodiscard]] T& operator*() const noexcept {
+            #if SANITIZE_BUMP_ALLOCATOR
+                EXPENSIVE_ASSERT(_size);
+            #endif
+            return get()[0];
+        };
+        [[nodiscard]] T* operator->() const noexcept {
+            #if SANITIZE_BUMP_ALLOCATOR
+                EXPENSIVE_ASSERT(_size);
+            #endif
+            return get();
+        };
+        // Returns a pointer which represents the tail of this allocation,
+        // but starting at `n`th element.
+        [[nodiscard]] ptr offset(size_t n) const noexcept {
+            auto ret = *this;
+            ret._ptr += n;
+            #if SANITIZE_BUMP_ALLOCATOR
+                size_t sz = n * sizeof(T);
+                EXPENSIVE_ASSERT(ret._size >= sz);
+                ret._global_pos += sz;
+                ret._size -= sz;
+            #endif
+            return ret;
+        };
+        // Splits the allocation in two.
+        [[nodiscard]] std::pair<ptr, ptr> split(size_t first_size) const noexcept {
+            auto ret1 = *this;
+            #if SANITIZE_BUMP_ALLOCATOR
+                size_t sz = first_size * sizeof(T);
+                EXPENSIVE_ASSERT(_size > sz);
+                ret1._size = sz;
+            #endif
+            auto ret2 = this->offset(first_size);
+            return {ret1, ret2};
+        };
+    };
+
+    constexpr static size_t segment_alignment = alignof(max_align_t);
+    // As usual in Scylla, the allocation stack is fragmented to avoid large contiguous buffers.
+    size_t _segment_size;
+
+private:
+    using buf = std::unique_ptr<std::byte[]>;
+    // We keep the last freed segment here. If we need a new buffer and this is present, we reuse it instead of allocating a new one.
+    // This ensures we won't repeatedly free and allocate a segment from the main allocator.
+    buf _spare = nullptr;
+    // The segment currently at the top of the stack.
+    buf _current = nullptr;
+    // Remaining free space in _current.
+    size_t _remaining = 0;
+    // Previous segments in the stack.
+    std::vector<buf> _old;
+
+#if SANITIZE_BUMP_ALLOCATOR
+    // The number of discard() calls until now.
+    uint64_t _discard_timestamp = 0;
```

Comment:
If it's a counter and not a TIMEstamp, don't call it a timestamp.

Reply:
Renamed from `_discard_timestamp` to `_discard_counter`.